# What are the type of apps attract more users?

## 1. Introduction:
This project is to analyze data to help our developers understand what type of apps are likely to attract more users.

## 2. Exploring the dataset content:

let's open and start to explore the dataset

In [1]:
def open_data(file_name):
    file_opened = open(file_name)
    from csv import reader
    file_reader = reader(file_opened)
    data_set = list(file_reader)
    return data_set

In [2]:
apple_store_data = open_data('AppleStore.csv')


In [3]:
google_play_data = open_data('googleplaystore.csv')

In [4]:
def explore_data(dataset, start, end, rows_and_columns=False):
    dataset_slice = dataset[start:end]
    for row in dataset_slice:
        print(row)
        print('\n')
        
    if rows_and_columns:
        print('Number of rows: ', len(dataset))
        print('number of columns: ', len(dataset[0]))
    

In [5]:
explore_data(apple_store_data, 1, 3, True)
explore_data(google_play_data, 1, 3, True)

['284882215', 'Facebook', '389879808', 'USD', '0.0', '2974676', '212', '3.5', '3.5', '95.0', '4+', 'Social Networking', '37', '1', '29', '1']


['389801252', 'Instagram', '113954816', 'USD', '0.0', '2161558', '1289', '4.5', '4.0', '10.23', '12+', 'Photo & Video', '37', '0', '29', '1']


Number of rows:  7198
number of columns:  16
['Photo Editor & Candy Camera & Grid & ScrapBook', 'ART_AND_DESIGN', '4.1', '159', '19M', '10,000+', 'Free', '0', 'Everyone', 'Art & Design', 'January 7, 2018', '1.0.0', '4.0.3 and up']


['Coloring book moana', 'ART_AND_DESIGN', '3.9', '967', '14M', '500,000+', 'Free', '0', 'Everyone', 'Art & Design;Pretend Play', 'January 15, 2018', '2.0.0', '4.0.3 and up']


Number of rows:  10842
number of columns:  13


column names: 

In [6]:
explore_data(apple_store_data, 0, 1)
explore_data(google_play_data, 0, 1)

['id', 'track_name', 'size_bytes', 'currency', 'price', 'rating_count_tot', 'rating_count_ver', 'user_rating', 'user_rating_ver', 'ver', 'cont_rating', 'prime_genre', 'sup_devices.num', 'ipadSc_urls.num', 'lang.num', 'vpp_lic']


['App', 'Category', 'Rating', 'Reviews', 'Size', 'Installs', 'Type', 'Price', 'Content Rating', 'Genres', 'Last Updated', 'Current Ver', 'Android Ver']




In [7]:
android_header = google_play_data[0]
print(android_header)

['App', 'Category', 'Rating', 'Reviews', 'Size', 'Installs', 'Type', 'Price', 'Content Rating', 'Genres', 'Last Updated', 'Current Ver', 'Android Ver']


## 3. Data Cleaning

### 3.1 Removing Duplicate:


*Here we try to find data missing:*
The answer is in row 10473 there is missing data in catagory column. we find that by notice the rating value which is incurrect.  

In [8]:
for row in google_play_data[1:]:
    if len(row) != len(android_header):
        print(row)
        print("\n")
        print("Index postion is:", google_play_data.index(row))

['Life Made WI-Fi Touchscreen Photo Frame', '1.9', '19', '3.0M', '1,000+', 'Free', '0', 'Everyone', '', 'February 11, 2018', '1.0.19', '4.0 and up']


Index postion is: 10473


In [9]:
print(google_play_data[10473])

['Life Made WI-Fi Touchscreen Photo Frame', '1.9', '19', '3.0M', '1,000+', 'Free', '0', 'Everyone', '', 'February 11, 2018', '1.0.19', '4.0 and up']


Lets delete the row 10473:

In [10]:
del google_play_data[10473]

 Now we want to check if the "google play" dataset has duplicate:

In [11]:
unique_value=[]
duplicated_value=[]

for app in google_play_data:
    if app[0] in unique_value:
        duplicated_value.append(app[0])
    else:
        unique_value.append(app[0])
        
print('Number of duplicate apps: ', len(duplicated_value))
print()
print('Examples of duplicate apps: \n', duplicated_value[:15])


Number of duplicate apps:  1181

Examples of duplicate apps: 
 ['Quick PDF Scanner + OCR FREE', 'Box', 'Google My Business', 'ZOOM Cloud Meetings', 'join.me - Simple Meetings', 'Box', 'Zenefits', 'Google Ads', 'Google My Business', 'Slack', 'FreshBooks Classic', 'Insightly CRM', 'QuickBooks Accounting: Invoicing & Expenses', 'HipChat - Chat Built for Teams', 'Xero Accounting Software']


**For example, the "Box" app has duplicates:**

In [12]:
for app in google_play_data:
    if app[0] == 'Box':
        print(app)

['Box', 'BUSINESS', '4.2', '159872', 'Varies with device', '10,000,000+', 'Free', '0', 'Everyone', 'Business', 'July 31, 2018', 'Varies with device', 'Varies with device']
['Box', 'BUSINESS', '4.2', '159872', 'Varies with device', '10,000,000+', 'Free', '0', 'Everyone', 'Business', 'July 31, 2018', 'Varies with device', 'Varies with device']
['Box', 'BUSINESS', '4.2', '159872', 'Varies with device', '10,000,000+', 'Free', '0', 'Everyone', 'Business', 'July 31, 2018', 'Varies with device', 'Varies with device']


**Another example, let's see the "Instagram" app:**

In [13]:
for app in google_play_data:
    if app[0] == 'Instagram':
        print(app)

['Instagram', 'SOCIAL', '4.5', '66577313', 'Varies with device', '1,000,000,000+', 'Free', '0', 'Teen', 'Social', 'July 31, 2018', 'Varies with device', 'Varies with device']
['Instagram', 'SOCIAL', '4.5', '66577446', 'Varies with device', '1,000,000,000+', 'Free', '0', 'Teen', 'Social', 'July 31, 2018', 'Varies with device', 'Varies with device']
['Instagram', 'SOCIAL', '4.5', '66577313', 'Varies with device', '1,000,000,000+', 'Free', '0', 'Teen', 'Social', 'July 31, 2018', 'Varies with device', 'Varies with device']
['Instagram', 'SOCIAL', '4.5', '66509917', 'Varies with device', '1,000,000,000+', 'Free', '0', 'Teen', 'Social', 'July 31, 2018', 'Varies with device', 'Varies with device']


**Notice that the 4th column has different values. The different numbers show the data was collected at different times. so we will keep the more recent**

In [14]:
print(google_play_data[10473])

['osmino Wi-Fi: free WiFi', 'TOOLS', '4.2', '134203', '4.1M', '10,000,000+', 'Free', '0', 'Everyone', 'Tools', 'August 7, 2018', '6.06.14', '4.4 and up']


Now, we will remove duplicates:

In [15]:
reviews_max = {}

for row in google_play_data[1:]:
    name=row[0]
    reviews = float(row[3])
       
    if name in reviews_max and reviews_max[name]<reviews:
        reviews_max[name]=reviews
    
    elif name not in reviews_max:
        reviews_max[name]=reviews


**In a previous code cell, we found that there are 1,181 cases where an app occurs more than once, so the length of our dictionary (of unique apps) should be equal to the difference between the length of our data set and 1,181:**

In [16]:
print('Expected length:', len(google_play_data[1:]) - 1181)
print('Actual length:', len(reviews_max))

Expected length: 9659
Actual length: 9659


Now, we update our dataset to google play data cleaned, we do that by 2 new lists: 

In [17]:
google_play_data_clean = []
already_added = []

for row in google_play_data[1:]:
    name = row[0]
    reviews = float(row[3])
    
    if reviews == reviews_max[name] and name not in already_added:
        google_play_data_clean.append(row)
        already_added.append(name)
        

lets check our data, to ensure everything went as expected:

In [18]:
print(len(google_play_data_clean))

9659


### 3.2 Removing Non-English Apps:

here we create a function that check non english apps:

In [19]:
def check_eng(a_string):
    characters_fallen = 0
    
    for character in a_string:
        if ord(character) > 127:
            characters_fallen+=1
            
    if characters_fallen > 3:
        return False
    else:
        return True

In [20]:
print(check_eng('Docs To Go™ Free Office Suite'))
print(check_eng('Instachat 😜'))
print(check_eng('爱奇艺PPS -《欢乐颂2》电视剧热播'))

True
True
False


Now we filter out non-English apps:

In [21]:
google_play_data_clean1=[]

for row in google_play_data_clean:
    name=row[0]
    if check_eng(name):
        google_play_data_clean1.append(row)
    

In [22]:
apple_store_data_clean=[]

for row in apple_store_data:
    name=row[1]
    if check_eng(name):
        apple_store_data_clean.append(row)


In [23]:
print('Google Data rows')
print('Before:', len(apple_store_data))
print('After:', len(apple_store_data_clean))
print()
print('App store rows')
print('Before:', len(google_play_data_clean))
print('After:', len(google_play_data_clean1))

Google Data rows
Before: 7198
After: 6184

App store rows
Before: 9659
After: 9614


### 3.3 Isolating the free apps:

**Here our updated datasets**:
apple_store_data_clean, 
google_play_data_clean1

In [24]:
google_play_data_clean2 = []

for row in google_play_data_clean1:
    price = row[6]
    
    if price == 'Free':
        google_play_data_clean2.append(row)

In [25]:
print(len(google_play_data_clean2))
print(len(google_play_data_clean1))

8863
9614


In [26]:
apple_store_data_clean1 =[]

for row in apple_store_data_clean[1:]:
    price = float(row[4])
    
    if price == 0.0:
        apple_store_data_clean1.append(row)

In [27]:
print(len(apple_store_data_clean1))
print(len(apple_store_data_clean))

3222
6184


## 4. Most Common Apps by Genre

As we mentioned in the introduction, our aim is to determine the kinds of apps that are likely to attract more users because our revenue is highly influenced by the number of people using our apps.

To minimize risks and overhead, our validation strategy for an app idea is comprised of three steps:

1. Build a minimal Android version of the app, and add it to Google Play.
2. If the app has a good response from users, we then develop it further.
3. If the app is profitable after six months, we also build an iOS version of the app and add it to the App Store.

Because our end goal is to add the app on both the App Store and Google Play, we need to find app profiles that are successful on both markets. For instance, a profile that might work well for both markets might be a productivity app that makes use of gamification.

Let's begin the analysis by getting a sense of the most common genres for each market. For this, we'll build a frequency table for the prime_genre column of the App Store data set, and the Genres and Category columns of the Google Play data set.

**The updated datasets:** google_play_data_clean2, apple_store_data_clean1

finding the relevante column:

In [28]:
for row in google_play_data_clean2:
    print(row[9])

Art & Design
Art & Design
Art & Design
Art & Design;Creativity
Art & Design
Art & Design
Art & Design
Art & Design
Art & Design;Creativity
Art & Design
Art & Design
Art & Design
Art & Design
Art & Design
Art & Design
Art & Design
Art & Design
Art & Design
Art & Design
Art & Design
Art & Design
Art & Design
Art & Design
Art & Design
Art & Design;Creativity
Art & Design
Art & Design
Art & Design
Art & Design
Art & Design
Art & Design
Art & Design
Art & Design
Art & Design
Art & Design
Art & Design
Art & Design
Art & Design
Art & Design
Art & Design;Creativity
Art & Design
Art & Design
Art & Design
Art & Design
Art & Design
Auto & Vehicles
Auto & Vehicles
Auto & Vehicles
Auto & Vehicles
Auto & Vehicles
Auto & Vehicles
Auto & Vehicles
Auto & Vehicles
Auto & Vehicles
Auto & Vehicles
Auto & Vehicles
Auto & Vehicles
Auto & Vehicles
Auto & Vehicles
Auto & Vehicles
Auto & Vehicles
Auto & Vehicles
Auto & Vehicles
Auto & Vehicles
Auto & Vehicles
Auto & Vehicles
Auto & Vehicles
Auto & Vehicles
Aut

finding the relevante column:

In [29]:
for row in apple_store_data_clean1:
    print(row[11])


Social Networking
Photo & Video
Games
Games
Music
Social Networking
Reference
Games
Music
Games
Games
Games
Games
Games
Games
Games
Games
Health & Fitness
Games
Weather
Games
Utilities
Games
Games
Travel
Games
Shopping
Games
Games
Games
Games
Music
Games
Games
Games
Games
Games
Health & Fitness
Social Networking
Games
Games
News
Social Networking
Navigation
Lifestyle
Social Networking
Games
Photo & Video
Entertainment
Food & Drink
Games
Games
Games
Music
Entertainment
Sports
Social Networking
Games
Photo & Video
Games
Games
Games
Games
Shopping
Social Networking
Food & Drink
Games
Games
Games
Book
Entertainment
Games
Games
Finance
Finance
Games
Travel
Games
Games
Weather
Reference
Games
Games
Games
Sports
Weather
Games
Games
Games
Games
Games
Games
Games
Entertainment
Games
Games
Games
Games
Games
Games
Social Networking
Games
Games
Entertainment
Games
Games
Social Networking
Games
Games
Social Networking
Education
Productivity
Games
Sports
Games
Games
Games
Games
Navigation
Games
Weat

freq_table function & display_table:

In [30]:
def freq_table(dataset, index):
    table = {}
    total = 0
    
    for row in dataset:
        total += 1
        value = row[index]
        if value in table:
            table[value] += 1
        else:
            table[value] = 1
            
    table_percentages = {}
    for key in table:
        percentage = (table[key] / total) * 100
        table_percentages[key] = percentage 
    
    return table_percentages


def display_table(dataset, index):
    table = freq_table(dataset, index)
    table_display = []
    for key in table:
        key_val_as_tuple = (table[key], key)
        table_display.append(key_val_as_tuple)
        
    table_sorted = sorted(table_display, reverse = True)
    for entry in table_sorted:
        print(entry[1], ':', entry[0])

Now we will use display_table function to analyze the datasets:

In [31]:
display_table(apple_store_data_clean1, 11)

Games : 58.16263190564867
Entertainment : 7.883302296710118
Photo & Video : 4.9658597144630665
Education : 3.662321539416512
Social Networking : 3.2898820608317814
Shopping : 2.60707635009311
Utilities : 2.5139664804469275
Sports : 2.1415270018621975
Music : 2.0484171322160147
Health & Fitness : 2.0173805090006205
Productivity : 1.7380509000620732
Lifestyle : 1.5828677839851024
News : 1.3345747982619491
Travel : 1.2414649286157666
Finance : 1.1173184357541899
Weather : 0.8690254500310366
Food & Drink : 0.8069522036002483
Reference : 0.5586592178770949
Business : 0.5276225946617008
Book : 0.4345127250155183
Navigation : 0.186219739292365
Medical : 0.186219739292365
Catalogs : 0.12414649286157665


**Summary Analysis:**
1. the most common genre is: 'Games'
2. the next most common is: 'Entertainment'
3. the general impression is that the most of the apps designed more for fun and entertainment.
4. However, the fact that fun apps are the most numerous doesn't also imply that they also have the greatest number of users — the demand might not be the same as the offer.

In [32]:
display_table(google_play_data_clean2, 1)

FAMILY : 18.898792733837304
GAME : 9.725826469592688
TOOLS : 8.462146000225657
BUSINESS : 4.592124562789123
LIFESTYLE : 3.9038700214374367
PRODUCTIVITY : 3.8925871601038025
FINANCE : 3.7007785174320205
MEDICAL : 3.5315355974275078
SPORTS : 3.396141261423897
PERSONALIZATION : 3.317161232088458
COMMUNICATION : 3.2381812027530184
HEALTH_AND_FITNESS : 3.0802211440821394
PHOTOGRAPHY : 2.944826808078529
NEWS_AND_MAGAZINES : 2.798149610741284
SOCIAL : 2.6627552747376737
TRAVEL_AND_LOCAL : 2.335552296062281
SHOPPING : 2.245289405393208
BOOKS_AND_REFERENCE : 2.1437436533904997
DATING : 1.8616721200496444
VIDEO_PLAYERS : 1.7939749520478394
MAPS_AND_NAVIGATION : 1.399074805370642
FOOD_AND_DRINK : 1.241114746699763
EDUCATION : 1.1621347173643235
ENTERTAINMENT : 0.9590432133589079
LIBRARIES_AND_DEMO : 0.9364774906916393
AUTO_AND_VEHICLES : 0.9251946293580051
HOUSE_AND_HOME : 0.8236488773552973
WEATHER : 0.8010831546880289
EVENTS : 0.7108202640189552
PARENTING : 0.6544059573507841
ART_AND_DESIGN : 0

In [33]:
display_table(google_play_data_clean2, 9)

Tools : 8.450863138892023
Entertainment : 6.070179397495204
Education : 5.348076272142616
Business : 4.592124562789123
Productivity : 3.8925871601038025
Lifestyle : 3.8925871601038025
Finance : 3.7007785174320205
Medical : 3.5315355974275078
Sports : 3.463838429425702
Personalization : 3.317161232088458
Communication : 3.2381812027530184
Action : 3.102786866749408
Health & Fitness : 3.0802211440821394
Photography : 2.944826808078529
News & Magazines : 2.798149610741284
Social : 2.6627552747376737
Travel & Local : 2.324269434728647
Shopping : 2.245289405393208
Books & Reference : 2.1437436533904997
Simulation : 2.042197901387792
Dating : 1.8616721200496444
Arcade : 1.8503892587160102
Video Players & Editors : 1.771409229380571
Casual : 1.7601263680469368
Maps & Navigation : 1.399074805370642
Food & Drink : 1.241114746699763
Puzzle : 1.128286133363421
Racing : 0.9928917973598104
Role Playing : 0.9364774906916393
Libraries & Demo : 0.9364774906916393
Auto & Vehicles : 0.9251946293580051
S

**Summary Analysis:**
1. the most common genres: 'Tools'
2. followed by 'Entertainment' apps and then Education apps.
3. compared to the previous analysis, we can see here that doesn't happen the same, the fun apps not is the most popular in this market, and it seems that a good number of apps are designed for practical purposes (family, tools, business, lifestyle, productivity, etc.). However, if we investigate this further, we can see that the family category (which accounts for almost 19% of the apps) means mostly games for kids

The difference between the Genres and the Category columns is not crystal clear, but one thing we can notice is that the Genres column is much more granular (it has more categories). We're only looking for the bigger picture at the moment, so we'll only work with the Category column moving forward.

Up to this point, we found that the App Store is dominated by apps designed for fun, while Google Play shows a more balanced landscape of both practical and for-fun apps. Now we'd like to get an idea about the kind of apps that have most users.

In [34]:
table = freq_table(apple_store_data_clean1, 11)

for genre in table:
    total=0
    len_genre=0
    
    for row in apple_store_data_clean1:
        genre_app = row[11]
        
        if genre_app == genre:
            n_user_ratings=float(row[5])
            total += n_user_ratings
            len_genre += 1
            
    avg_n_user_ratings = total/len_genre
    print(genre,' : ',avg_n_user_ratings)

Social Networking  :  71548.34905660378
Photo & Video  :  28441.54375
Games  :  22788.6696905016
Music  :  57326.530303030304
Reference  :  74942.11111111111
Health & Fitness  :  23298.015384615384
Weather  :  52279.892857142855
Utilities  :  18684.456790123455
Travel  :  28243.8
Shopping  :  26919.690476190477
News  :  21248.023255813954
Navigation  :  86090.33333333333
Lifestyle  :  16485.764705882353
Entertainment  :  14029.830708661417
Food & Drink  :  33333.92307692308
Sports  :  23008.898550724636
Book  :  39758.5
Finance  :  31467.944444444445
Education  :  7003.983050847458
Productivity  :  21028.410714285714
Business  :  7491.117647058823
Catalogs  :  4004.0
Medical  :  612.0


On average, navigation apps have the highest number of user reviews, but this figure is heavily influenced by Waze and Google Maps, which have close to half a million user reviews together:

In [35]:
for row in apple_store_data_clean1:
    if row[-5] == 'Navigation':
        print(row[1], ':', row[5]) # print name and number of ratings

Waze - GPS Navigation, Maps & Real-time Traffic : 345046
Google Maps - Navigation & Transit : 154911
Geocaching® : 12811
CoPilot GPS – Car Navigation & Offline Maps : 3582
ImmobilienScout24: Real Estate Search in Germany : 187
Railway Route Search : 5


### 4.2 Most Popular Apps by Genre on Google Play:

In [40]:
freq_table(google_play_data_clean2, 1)

{'ART_AND_DESIGN': 0.6431230960171499,
 'AUTO_AND_VEHICLES': 0.9251946293580051,
 'BEAUTY': 0.5979916506826132,
 'BOOKS_AND_REFERENCE': 2.1437436533904997,
 'BUSINESS': 4.592124562789123,
 'COMICS': 0.6205573733498815,
 'COMMUNICATION': 3.2381812027530184,
 'DATING': 1.8616721200496444,
 'EDUCATION': 1.1621347173643235,
 'ENTERTAINMENT': 0.9590432133589079,
 'EVENTS': 0.7108202640189552,
 'FINANCE': 3.7007785174320205,
 'FOOD_AND_DRINK': 1.241114746699763,
 'HEALTH_AND_FITNESS': 3.0802211440821394,
 'HOUSE_AND_HOME': 0.8236488773552973,
 'LIBRARIES_AND_DEMO': 0.9364774906916393,
 'LIFESTYLE': 3.9038700214374367,
 'GAME': 9.725826469592688,
 'FAMILY': 18.898792733837304,
 'MEDICAL': 3.5315355974275078,
 'SOCIAL': 2.6627552747376737,
 'SHOPPING': 2.245289405393208,
 'PHOTOGRAPHY': 2.944826808078529,
 'SPORTS': 3.396141261423897,
 'TRAVEL_AND_LOCAL': 2.335552296062281,
 'TOOLS': 8.462146000225657,
 'PERSONALIZATION': 3.317161232088458,
 'PRODUCTIVITY': 3.8925871601038025,
 'PARENTING': 0.

In [48]:
categories_android = freq_table(google_play_data_clean2, 1)

for category in categories_android:
    total = 0
    len_category = 0
    for app in google_play_data_clean2:
        category_app = app[1]
        if category_app == category:            
            n_installs = app[5]
            n_installs = n_installs.replace(',', '')
            n_installs = n_installs.replace('+', '')
            total += float(n_installs)
            len_category += 1
    avg_n_installs = total / len_category
    print(category, ':', avg_n_installs)

ART_AND_DESIGN : 1986335.0877192982
AUTO_AND_VEHICLES : 647317.8170731707
BEAUTY : 513151.88679245283
BOOKS_AND_REFERENCE : 8767811.894736841
BUSINESS : 1712290.1474201474
COMICS : 817657.2727272727
COMMUNICATION : 38456119.167247385
DATING : 854028.8303030303
EDUCATION : 1833495.145631068
ENTERTAINMENT : 11640705.88235294
EVENTS : 253542.22222222222
FINANCE : 1387692.475609756
FOOD_AND_DRINK : 1924897.7363636363
HEALTH_AND_FITNESS : 4188821.9853479853
HOUSE_AND_HOME : 1331540.5616438356
LIBRARIES_AND_DEMO : 638503.734939759
LIFESTYLE : 1437816.2687861272
GAME : 15588015.603248259
FAMILY : 3697848.1731343283
MEDICAL : 120550.61980830671
SOCIAL : 23253652.127118643
SHOPPING : 7036877.311557789
PHOTOGRAPHY : 17840110.40229885
SPORTS : 3638640.1428571427
TRAVEL_AND_LOCAL : 13984077.710144928
TOOLS : 10801391.298666667
PERSONALIZATION : 5201482.6122448975
PRODUCTIVITY : 16787331.344927534
PARENTING : 542603.6206896552
WEATHER : 5074486.197183099
VIDEO_PLAYERS : 24727872.452830188
NEWS_AND_

On average, communication apps have the most installs: 38,456,119.

### 5. Conclusions
In this project, we analyzed data about the App Store and Google Play mobile apps with the goal of recommending an app profile that can be profitable for both markets.

We concluded that taking a popular book (perhaps a more recent book) and turning it into an app could be profitable for both the Google Play and the App Store markets. The markets are already full of libraries, so we need to add some special features besides the raw version of the book. This might include daily quotes from the book, an audio version of the book, quizzes on the book, a forum where people can discuss the book, etc.